src/Components/Components/src/NavigationManagerExtensions.cs

```diff
@@ -0,0 +1,938 @@
+        /// <param name="values">The parameter values to add or update.</param>
+        /// <remarks>
+        /// Any <c>null</c> entries in <paramref name="values"/> will be skipped. Existing querystring parameters not in
+        /// <paramref name="values"/> will be removed from the querystring in the returned URI.
+        /// </remarks>
+        public static string UriWithQueryParameter(this NavigationManager navigationManager, string name, IEnumerable<long?> values)
+            => UriWithQueryParameter(navigationManager, name, values, Format);
+
+        /// <summary>
+        /// Returns a URI that is constructed by updating <see cref="NavigationManager.Uri"/> with a single parameter
+        /// added, updated, or removed.
+        /// </summary>
+        /// <param name="navigationManager">The <see cref="NavigationManager"/>.</param>
+        /// <param name="name">The name of the parameter to add or update.</param>
+        /// <param name="value">The value of the parameter to add or update.</param>
+        /// <remarks>
+        /// If <paramref name="value"/> is <c>null</c>, the parameter will be removed if it exists in the URI.
+        /// Otherwise, it will be added or updated.
+        /// </remarks>
+        public static string UriWithQueryParameter(this NavigationManager navigationManager, string name, string? value)
+        {
+            if (navigationManager is null)
+            {
+                throw new ArgumentNullException(nameof(navigationManager));
+            }
+
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new InvalidOperationException(EmptyQueryParameterNameExceptionMessage);
+            }
+
+            var uri = navigationManager.Uri;
+
+            return value is null
+                ? UriWithoutQueryParameter(uri, name)
+                : UriWithQueryParameterCore(uri, name, value);
+        }
+
+        private static string UriWithQueryParameter<TValue>(
+            this NavigationManager navigationManager,
+            string name,
+            IEnumerable<TValue?> values,
+            QueryParameterFormatter<TValue> formatter)
+        {
+            if (navigationManager is null)
+            {
+                throw new ArgumentNullException(nameof(navigationManager));
+            }
+
+            var uri = navigationManager.Uri;
+            var source = new QueryParameterSource<TValue>(name, values, formatter);
+
+            if (!TryRebuildExistingQueryFromUri(uri, out var existingQueryStringEnumerable, out var newQueryStringBuilder))
+            {
+                return UriWithAppendedQueryParameters(uri, ref source);
+            }
+
+            foreach (var pair in existingQueryStringEnumerable)
+            {
+                if (pair.EncodedName.Span.Equals(source.EncodedName, StringComparison.OrdinalIgnoreCase))
+                {
+                    source.AppendNextParameter(ref newQueryStringBuilder);
```

Comment:
This confused me for a bit because it looked as if it was going to append another copy of the value every time it saw an earlier query param with the same name.

Am I right in understanding it only emits one copy because `AppendNextParameter` knows to only do it once? If so this is cool and efficient, but might warrant adding a comment above L725 to clarify.

Reply:
Yes, it only emits one copy. I will add a comment to clarify this behavior. Maybe a better name than `AppendNextParameter` could also imply that it works this way.